python/lsst/daf/butler/core/datastore.py

```diff
@@ -85,6 +94,19 @@ class Event:
     kwargs: dict
 
 
+@dataclasses.dataclass(frozen=True)
+class OpaqueTableDefinition:
+    """Definition of the opaque table which stores datastore records.
+
+    Table definition contains `.ddl.TableSpec` for a table and a class
+    of a record which must be a subclass of `StoredDatastoreItemInfo`.
+    """
+
+    __slots__ = {"table_spec", "record_class"}
+    table_spec: ddl.TableSpec
+    record_class: type[StoredDatastoreItemInfo]
```

Comment:
This is really a minor side issue, but I'd prefer not to assume anywhere that opaque tables are just for datastores, and I'd be great to cut down on the number of conversions (SQLAlchemy row <-> `StoredDatastoreItemInfo` <-> pydantic/json mapping).

So, could we put this `OpaqueTableDefinition` struct in `registry.interfaces.opaque`?

And could we make `record_class` just "some pydantic model type"?  Or maybe even just use bare `dict` here?   Dynamic class types aren't getting us any static-analysis advantages, and runtime checking is probably a waste of cycles (except at the JSON boundary).

Reply:
I think it should be OK to move `OpaqueTableDefinition` to `registry.interfaces`. I was worried that this may cause `core` to be (logically) dependent on `registry`, but we can do it as mypy-only dependency, as we do for a couple of other classes already. We do need then to replace `StoredDatastoreItemInfo` with registry-side type (or possibly use `dict`), I need to look at the code and figure out a reasonable option.

Reply:
I don't like the idea of completely abandoning typing and using dict for arbitrary datastore records. The `StoredDatastoreItemInfo` dataclass definition is seemingly very close to defining the database schema directly through introspection.

Reply:
I would like to have typed datastore records, but it makes life a bit harder for registry. But it's probably unavoidable.